README.md
<div id="top"></div>

[![Contributors][contributors-shield]][contributors-url]
[![Forks][forks-shield]][forks-url]
[![Stargazers][stars-shield]][stars-url]
[![Issues][issues-shield]][issues-url]
[![MIT License][license-shield]][license-url]
[![LinkedIn][linkedin-shield]][linkedin-url]

<!-- PROJECT LOGO -->
<br />
<div align="center">
  <a href="https://github.com/kinshukdua/gUPI-recon">
    <img src="./logo.png" alt="Logo" width="250" height="250">
  </a>

<h2 align="center">gUPI-recon</h2>

  <p align="center">
A GUI for reconnaissance using virtual payment address (VPA).<br />
This tool leverages the openness available with the UPI platform to find :<br />
UPI ID and name associated with a mobile number.<br />
UPI ID and name associated with a gmail account.<br />
UPI ID and name associated with a vehicle registration number.<br />
Leveraging UPI id associated with a FASTag.<br />
This project is a GUI port of upi-recon by @squeal.<br />
    <br />
    <a href="https://github.com/kinshukdua/gUPI-recon"><strong>Explore the docs »</strong></a>
    <br />
    <br />
    <a href="https://github.com/kinshukdua/gUPI-recon">View Demo</a>
    ·
    <a href="https://github.com/kinshukdua/gUPI-recon/issues">Report Bug</a>
    ·
    <a href="https://github.com/kinshukdua/gUPI-recon/issues">Request Feature</a>
  </p>
</div>



<!-- TABLE OF CONTENTS -->
<details>
  <summary>Table of Contents</summary>
  <li>
    <li>
      <a href="#about-the-project">About The Project</a>
      <ul>
        <li><a href="#built-with">Built With</a></li>
      </ul>
    </li>
    <li>
      <a href="#getting-started">Getting Started</a>
      <ul>
        <li><a href="#requirements">Requirements</a></li>
      </ul>
    </li>
    <li><a href="#usage">Usage</a></li>
    <li><a href="#roadmap">Roadmap</a></li>
    <li><a href="#contributing">Contributing</a></li>
    <li><a href="#license">License</a></li>
    <li><a href="#contact">Contact</a></li>
    <li><a href="#disclaimer">Disclaimer</a></li>
  </li>
</details>



<!-- ABOUT THE PROJECT -->
## About The Project

[![Product Name Screen Shot][product-screenshot]]()



<p align="right">(<a href="#top">back to top</a>)</p>



### Built With

* [upi-recon](https://github.com/qurbat/upi-recon/)
* [PyQt5](https://pypi.org/project/PyQt5/)

<p align="right">(<a href="#top">back to top</a>)</p>



<!-- GETTING STARTED -->
## Getting Started

### Requirements

* Install all the requirments

  ```sh
  pip install -r requirements.txt
  ```

<p align="right">(<a href="#top">back to top</a>)</p>



<!-- USAGE EXAMPLES -->
## Usage
* STEP 1: Enter the mobile number or gmail address or vehicle registration number
     
[![Step 1][usage-1]]()
    
* STEP 2: Select the API to use and the UPI to Query
    
[![Step 2][usage-2]]()
   
* STEP 3: Click on the Query button
      
[![Step 3][usage-3]]()


<p align="right">(<a href="#top">back to top</a>)</p>



<!-- ROADMAP -->
## Roadmap
- [ ] Finish remaining features
- [ ] RazorPay API
- [ ] Custom API
- [ ] gmail recon


See the [open issues](https://github.com/kinshukdua/gUPI-recon/issues) for a full list of proposed features (and known issues).

<p align="right">(<a href="#top">back to top</a>)</p>



<!-- CONTRIBUTING -->
## Contributing

Contributions are what make the open source community such an amazing place to learn, inspire, and create. Any contributions you make are **greatly appreciated**.

If you have a suggestion that would make this better, please fork the repo and create a pull request. You can also simply open an issue with the tag "enhancement".
Don't forget to give the project a star! Thanks again!

1. Fork the Project
2. Create your Feature Branch (`git checkout -b feature/AmazingFeature`)
3. Commit your Changes (`git commit -m 'Add some AmazingFeature'`)
4. Push to the Branch (`git push origin feature/AmazingFeature`)
5. Open a Pull Request

<p align="right">(<a href="#top">back to top</a>)</p>



<!-- LICENSE -->
## License

Distributed under the MIT License. See `LICENSE.txt` for more information.

<p align="right">(<a href="#top">back to top</a>)</p>



<!-- CONTACT -->
## Contact

Kinshuk Dua - [@linkedIn](https://linkedin.com/in/dua)

Project Link: [https://github.com/kinshukdua/gUPI-recon](https://github.com/kinshukdua/gUPI-recon)

<p align="right">(<a href="#top">back to top</a>)</p>



<!-- ACKNOWLEDGMENTS -->
## Disclaimer
Note: Unified Payment Interface ("UPI") Virtual Payment Addresses ("VPAs") do not carry a data security classification by virtue of their usage in practice, and should as such be considered to be public information, similar to how email addresses may be considered to be public information.

This tool allows users to 1) check the existence of UPI payment addresses, and 2) fetch associated information about the account holder, in an automated manner based on provided input. This functionality is already available (however, not in an automated fashion) through most UPI payment applications available on the Android and/or iOS platforms.

This tool is provided "AS IS" without any warranty of any kind, either expressed, implied, or statutory, to the extent permitted by applicable law.
<p align="right">(<a href="#top">back to top</a>)</p>



<!-- MARKDOWN LINKS & IMAGES -->
<!-- https://www.markdownguide.org/basic-syntax/#reference-style-links -->
[contributors-shield]: https://img.shields.io/github/contributors/kinshukdua/gUPI-recon.svg?style=for-the-badge
[contributors-url]: https://github.com/kinshukdua/gUPI-recon/graphs/contributors
[forks-shield]: https://img.shields.io/github/forks/kinshukdua/gUPI-recon.svg?style=for-the-badge
[forks-url]: https://github.com/kinshukdua/gUPI-recon/network/members
[stars-shield]: https://img.shields.io/github/stars/kinshukdua/gUPI-recon.svg?style=for-the-badge
[stars-url]: https://github.com/kinshukdua/gUPI-recon/stargazers
[issues-shield]: https://img.shields.io/github/issues/kinshukdua/gUPI-recon.svg?style=for-the-badge
[issues-url]: https://github.com/kinshukdua/gUPI-recon/issues
[license-shield]: https://img.shields.io/github/license/kinshukdua/gUPI-recon.svg?style=for-the-badge
[license-url]: https://github.com/kinshukdua/gUPI-recon/blob/master/LICENSE.txt
[linkedin-shield]: https://img.shields.io/badge/-LinkedIn-black.svg?style=for-the-badge&logo=linkedin&colorB=555
[linkedin-url]: https://linkedin.com/in/dua
[product-screenshot]: screenshots/screenshot.png
[usage-1]: screenshots/usage1.png
[usage-2]: screenshots/usage2.PNG
[usage-3]: screenshots/usage3.PNG


CHANGELOG.md
# Changelog

## v0.0.1

### Added or Changed
- Initial release


app.py
import sys
import pyperclip 
from PyQt5.QtCore import QRegExp, QObject, QThread, pyqtSignal
from PyQt5.QtGui import QRegExpValidator, QIcon
from PyQt5.QtWidgets import (
    QApplication, QDialog, QMainWindow, QMessageBox, QDialogButtonBox, QLineEdit, QTableWidgetItem, QRadioButton, QFileDialog
)
from PyQt5 import QtWidgets
from PyQt5.uic import loadUi
from main_ui import Ui_MainWindow

import requests
import concurrent.futures

from time import sleep
from sys import exit
from random import uniform as rand
from datetime import datetime


API_URL = 'https://upibankvalidator.com/api/upiValidation?upi='

phone_regex = QRegExp("[0-9]{10}")
email_regex = QRegExp("^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\\.[a-zA-Z]{2,4}$")

#  opting to load lists from a file instead of hardcoding them
#  as this would be more flexible, allow for easier updates,
#  and allow others to make use of the lists provided
with open("data/general_suffixes.txt", "r") as suffix_file:
    upi_suffix_dict = suffix_file.read().splitlines() #  read all suffixes into a list

with open("data/mobile_suffixes.txt", "r") as mobile_suffix_file:
    mobile_suffix_dict = mobile_suffix_file.read().splitlines()

with open("data/fastag_suffixes.txt", "r") as fastag_suffix_file:
    fastag_suffix_dict = fastag_suffix_file.read().splitlines()

with open("data/gpay_suffixes.txt", "r") as gpay_suffix_file:
    gpay_suffix_dict = gpay_suffix_file.read().splitlines()



class Scraper(QObject):
    finished = pyqtSignal()
    progress = pyqtSignal(int)
    found = 0
    count = 0
    def searchvpa(self, searchtext, vpa_dict, threadcount):
        if(threadcount == 0):
            for suffix in vpa_dict:
                try:
                    self.address_discovery(searchtext + '@' + suffix, API_URL)
                except KeyboardInterrupt:
                    exit(0)
        else:
            threadcount = 10 if threadcount > 10 else threadcount
            with concurrent.futures.ThreadPoolExecutor(max_workers=threadcount) as executor:
                try:
                    for suffix in vpa_dict:
                        executor.submit(self.address_discovery, searchtext + '@' + suffix, API_URL)
                        sleep(rand(.1, .2))
                except KeyboardInterrupt:
                    #  quit ungracefully on keyboard interrupt:
                    #  considering the bandwidth consumed for requests,
                    #  there is no reason to wait for the threads to finish
                    #  sorry for the inconvenience
                    executor._threads.clear()
                    concurrent.futures.thread._threads_queues.clear()
        if self.found == 0:
            win.set_footer("No VPA found for: "+searchtext, "red")
        
    def address_discovery(self, vpa, api_url):
        r = requests.post(api_url+vpa)

        print(r.status_code)
        if r.status_code == 200 and r.json()['isUpiRegistered'] is True:
            rowPosition = win.Result.rowCount()
            win.Result.insertRow(rowPosition)
            win.Result.setItem(rowPosition , 0, QTableWidgetItem(r.json()['name']))
            win.Result.setItem(rowPosition , 1, QTableWidgetItem(vpa))
            win.nameEdit.setText(r.json()['name'])    
            self.found += 1
            win.set_footer("Found: "+ str(self.found), "green")
        if r.status_code == 400:
            win.set_footer("Bad Request", "red")
        if r.status_code == 429:
            win.set_footer("Too Many Requests", "red")
        self.progress.emit(self.count)
        self.count += 1
    def run(self):
        self.searchvpa(win.phoneEdit.text(), win.suffix, 2)
        self.finished.emit()

class Window(QMainWindow, Ui_MainWindow):
    def __init__(self, parent=None):
        super().__init__(parent)
        self.setupUi(self)
        self.connectSignalsSlots()
        self.phoneEdit.setValidator(QRegExpValidator(phone_regex, self.phoneEdit))
        self.emailEdit.setValidator(QRegExpValidator(email_regex, self.emailEdit))
        self.nameEdit.setReadOnly(True)
        self.progressBar.setVisible(False)
        self.allUPI.setChecked(True)
        self.upibankAPI.setChecked(True)
        self.Result.setColumnCount(2)  
        self.Result.horizontalHeader().setSectionResizeMode(QtWidgets.QHeaderView.Stretch)
        self.Result.setHorizontalHeaderLabels(["Name", "VPA"])
        self.setWindowIcon(QIcon('logo.png'))
        self.setWindowTitle("gUPI recon")

    def connectSignalsSlots(self):
        self.action_Manually.triggered.connect(self.add_api)
        self.pastePhone.clicked.connect(self.paste_phone)
        self.pasteEmail.clicked.connect(self.paste_email)
        self.Query.clicked.connect(self.query)
        self.saveButton.clicked.connect(self.saveQuery)
        self.action_Add_API_URL.triggered.connect(self.add_api)
        #self.action_Find_Replace.triggered.connect(self.findAndReplace)
        #self.action_About.triggered.connect(self.about)

    def add_api(self):
        pass

    def set_footer(self, text, color = "black"):
        self.footer.setStyleSheet("color: " + color)
        self.footer.setText(text)

    def add_api(self):
        dialog = add_API(self)
        dialog.exec()

    def paste_phone(self):
        if (QRegExp.exactMatch(phone_regex, pyperclip.paste())):
            self.phoneEdit.setText(pyperclip.paste())
        else:
            self.set_footer("Invalid phone number", "red")
    
    def paste_email(self):
        if (QRegExp.exactMatch(email_regex, pyperclip.paste())):
            self.emailEdit.setText(pyperclip.paste())
        else:
            self.set_footer("Invalid email", "red")

    def reportProgress(self, value):
        self.progressBar.setValue(round(((value+1)/len(self.suffix))*100))
        

    def saveQuery(self):
        if self.Result.rowCount() > 0:
            name, _ = QFileDialog.getSaveFileName(self, 'Save Query', self.phoneEdit.text(), 'CSV(*.csv)')
            if not name:
                return
            with open(name,'w') as file:
                for i in range(self.Result.rowCount()):
                    file.write(self.Result.item(i, 0).text())
                    file.write(", ")
                    file.write(self.Result.item(i, 1).text())
                    file.write("\n")
            footer_text = "Saved to " + name
            if len(footer_text) > 27:
                self.set_footer(footer_text[25:]+"...", "green")
            self.set_footer(footer_text, "green")
        else:
            self.set_footer("No results to save", "red")

    def query(self):
        # set footer
        self.set_footer("Querying...")
        self.thread = QThread()
        self.worker = Scraper()
        self.worker.moveToThread(self.thread)
        self.thread.started.connect(self.worker.run)
        self.worker.finished.connect(self.thread.quit)
        self.worker.finished.connect(self.worker.deleteLater)
        self.thread.finished.connect(self.thread.deleteLater)
        self.worker.progress.connect(self.reportProgress)
        self.progressBar.setValue(0)
        self.progressBar.setVisible(True)
        self.thread.finished.connect(
            lambda: self.progressBar.setVisible(False)
        )
        self.suffix = upi_suffix_dict
        if (win.gpayUPI.isChecked()):
            self.suffix = gpay_suffix_dict
        elif (win.fastagUPI.isChecked()):
            self.suffix = fastag_suffix_dict
        elif (win.mobileUPI.isChecked()):
            self.suffix = mobile_suffix_dict
        self.Result.clearContents()
        self.Result.setRowCount(0)
        self.thread.start()

class add_API(QDialog):
    def ok_clicked(self):
        print(self.api_input.text())
        self.close()
    def __init__(self, parent=None):
        super().__init__(parent)
        loadUi("UI/api_input.ui", self)
        self.api_input = self.findChild(QLineEdit, "lineEdit")
        self.button_box = self.findChild(QDialogButtonBox, "buttonBox")
        self.ok = self.button_box.buttons()[0]
        self.ok.clicked.connect(self.ok_clicked)


if __name__ == "__main__":
    app = QApplication(sys.argv)
    win = Window()
    win.show()
    sys.exit(app.exec())

main_ui.py
# -*- coding: utf-8 -*-

# Form implementation generated from reading ui file 'UI\main.ui'
#
# Created by: PyQt5 UI code generator 5.15.4
#
# WARNING: Any manual changes made to this file will be lost when pyuic5 is
# run again.  Do not edit this file unless you know what you are doing.


from PyQt5 import QtCore, QtGui, QtWidgets


class Ui_MainWindow(object):
    def setupUi(self, MainWindow):
        MainWindow.setObjectName("MainWindow")
        MainWindow.resize(351, 543)
        MainWindow.setMinimumSize(QtCore.QSize(351, 0))
        self.centralwidget = QtWidgets.QWidget(MainWindow)
        self.centralwidget.setObjectName("centralwidget")
        self.gridLayout = QtWidgets.QGridLayout(self.centralwidget)
        self.gridLayout.setObjectName("gridLayout")
        self.phoneEdit = QtWidgets.QLineEdit(self.centralwidget)
        self.phoneEdit.setObjectName("phoneEdit")
        self.gridLayout.addWidget(self.phoneEdit, 1, 0, 1, 2)
        self.nameEdit = QtWidgets.QLineEdit(self.centralwidget)
        self.nameEdit.setObjectName("nameEdit")
        self.gridLayout.addWidget(self.nameEdit, 17, 0, 2, 1)
        self.Query = QtWidgets.QPushButton(self.centralwidget)
        self.Query.setObjectName("Query")
        self.gridLayout.addWidget(self.Query, 17, 2, 1, 1)
        self.razorpayUPI = QtWidgets.QRadioButton(self.centralwidget)
        self.razorpayUPI.setObjectName("razorpayUPI")
        self.APIGroup = QtWidgets.QButtonGroup(MainWindow)
        self.APIGroup.setObjectName("APIGroup")
        self.APIGroup.addButton(self.razorpayUPI)
        self.gridLayout.addWidget(self.razorpayUPI, 11, 0, 1, 1)
        self.upibankAPI = QtWidgets.QRadioButton(self.centralwidget)
        self.upibankAPI.setObjectName("upibankAPI")
        self.APIGroup.addButton(self.upibankAPI)
        self.gridLayout.addWidget(self.upibankAPI, 8, 0, 1, 1)
        self.label_4 = QtWidgets.QLabel(self.centralwidget)
        font = QtGui.QFont()
        font.setPointSize(9)
        self.label_4.setFont(font)
        self.label_4.setObjectName("label_4")
        self.gridLayout.addWidget(self.label_4, 5, 0, 1, 1)
        self.line = QtWidgets.QFrame(self.centralwidget)
        self.line.setFrameShape(QtWidgets.QFrame.HLine)
        self.line.setFrameShadow(QtWidgets.QFrame.Sunken)
        self.line.setObjectName("line")
        self.gridLayout.addWidget(self.line, 22, 0, 1, 3)
        self.pasteEmail = QtWidgets.QPushButton(self.centralwidget)
        self.pasteEmail.setObjectName("pasteEmail")
        self.gridLayout.addWidget(self.pasteEmail, 6, 2, 1, 1)
        self.progressBar = QtWidgets.QProgressBar(self.centralwidget)
        self.progressBar.setProperty("value", 24)
        self.progressBar.setObjectName("progressBar")
        self.gridLayout.addWidget(self.progressBar, 20, 0, 1, 1)
        self.label = QtWidgets.QLabel(self.centralwidget)
        font = QtGui.QFont()
        font.setPointSize(9)
        self.label.setFont(font)
        self.label.setObjectName("label")
        self.gridLayout.addWidget(self.label, 16, 0, 1, 1)
        self.label_2 = QtWidgets.QLabel(self.centralwidget)
        font = QtGui.QFont()
        font.setPointSize(9)
        self.label_2.setFont(font)
        self.label_2.setObjectName("label_2")
        self.gridLayout.addWidget(self.label_2, 0, 0, 1, 1)
        self.allUPI = QtWidgets.QRadioButton(self.centralwidget)
        self.allUPI.setObjectName("allUPI")
        self.UPIGroup = QtWidgets.QButtonGroup(MainWindow)
        self.UPIGroup.setObjectName("UPIGroup")
        self.UPIGroup.addButton(self.allUPI)
        self.gridLayout.addWidget(self.allUPI, 16, 2, 1, 1)
        self.emailEdit = QtWidgets.QLineEdit(self.centralwidget)
        self.emailEdit.setObjectName("emailEdit")
        self.gridLayout.addWidget(self.emailEdit, 6, 0, 1, 2)
        self.mobileUPI = QtWidgets.QRadioButton(self.centralwidget)
        self.mobileUPI.setObjectName("mobileUPI")
        self.UPIGroup.addButton(self.mobileUPI)
        self.gridLayout.addWidget(self.mobileUPI, 8, 2, 1, 1)
        self.gpayUPI = QtWidgets.QRadioButton(self.centralwidget)
        self.gpayUPI.setObjectName("gpayUPI")
        self.UPIGroup.addButton(self.gpayUPI)
        self.gridLayout.addWidget(self.gpayUPI, 11, 2, 1, 1)
        self.saveButton = QtWidgets.QPushButton(self.centralwidget)
        self.saveButton.setObjectName("saveButton")
        self.gridLayout.addWidget(self.saveButton, 20, 2, 1, 1)
        self.Result = QtWidgets.QTableWidget(self.centralwidget)
        self.Result.setMinimumSize(QtCore.QSize(321, 0))
        self.Result.setObjectName("Result")
        self.Result.setColumnCount(0)
        self.Result.setRowCount(0)
        self.gridLayout.addWidget(self.Result, 19, 0, 1, 3)
        self.pastePhone = QtWidgets.QPushButton(self.centralwidget)
        self.pastePhone.setObjectName("pastePhone")
        self.gridLayout.addWidget(self.pastePhone, 1, 2, 1, 1)
        self.footer = QtWidgets.QLabel(self.centralwidget)
        font = QtGui.QFont()
        font.setPointSize(8)
        self.footer.setFont(font)
        self.footer.setObjectName("footer")
        self.gridLayout.addWidget(self.footer, 24, 0, 1, 1)
        self.fastagUPI = QtWidgets.QRadioButton(self.centralwidget)
        self.fastagUPI.setObjectName("fastagUPI")
        self.UPIGroup.addButton(self.fastagUPI)
        self.gridLayout.addWidget(self.fastagUPI, 12, 2, 1, 1)
        self.customAPI = QtWidgets.QRadioButton(self.centralwidget)
        self.customAPI.setObjectName("customAPI")
        self.APIGroup.addButton(self.customAPI)
        self.gridLayout.addWidget(self.customAPI, 12, 0, 1, 1)
        self.label.raise_()
        self.phoneEdit.raise_()
        self.saveButton.raise_()
        self.label_2.raise_()
        self.emailEdit.raise_()
        self.Query.raise_()
        self.label_4.raise_()
        self.pastePhone.raise_()
        self.pasteEmail.raise_()
        self.line.raise_()
        self.footer.raise_()
        self.Result.raise_()
        self.nameEdit.raise_()
        self.progressBar.raise_()
        self.mobileUPI.raise_()
        self.gpayUPI.raise_()
        self.fastagUPI.raise_()
        self.allUPI.raise_()
        self.upibankAPI.raise_()
        self.razorpayUPI.raise_()
        self.customAPI.raise_()
        MainWindow.setCentralWidget(self.centralwidget)
        self.menubar = QtWidgets.QMenuBar(MainWindow)
        self.menubar.setGeometry(QtCore.QRect(0, 0, 351, 26))
        self.menubar.setObjectName("menubar")
        self.menu_Open = QtWidgets.QMenu(self.menubar)
        self.menu_Open.setObjectName("menu_Open")
        self.menu_Add_API_2 = QtWidgets.QMenu(self.menu_Open)
        self.menu_Add_API_2.setObjectName("menu_Add_API_2")
        self.menu_Query = QtWidgets.QMenu(self.menubar)
        self.menu_Query.setObjectName("menu_Query")
        self.menu_About = QtWidgets.QMenu(self.menubar)
        self.menu_About.setObjectName("menu_About")
        MainWindow.setMenuBar(self.menubar)
        self.statusbar = QtWidgets.QStatusBar(MainWindow)
        self.statusbar.setObjectName("statusbar")
        MainWindow.setStatusBar(self.statusbar)
        self.actionManually = QtWidgets.QAction(MainWindow)
        self.actionManually.setObjectName("actionManually")
        self.action_Manually = QtWidgets.QAction(MainWindow)
        self.action_Manually.setObjectName("action_Manually")
        self.action_From_File = QtWidgets.QAction(MainWindow)
        self.action_From_File.setObjectName("action_From_File")
        self.action_New_Query = QtWidgets.QAction(MainWindow)
        self.action_New_Query.setObjectName("action_New_Query")
        self.actionSave_Query_Results = QtWidgets.QAction(MainWindow)
        self.actionSave_Query_Results.setObjectName("actionSave_Query_Results")
        self.action_Add_API_URL = QtWidgets.QAction(MainWindow)
        self.action_Add_API_URL.setObjectName("action_Add_API_URL")
        self.menu_Add_API_2.addAction(self.action_Manually)
        self.menu_Add_API_2.addAction(self.action_From_File)
        self.menu_Open.addAction(self.menu_Add_API_2.menuAction())
        self.menu_Open.addAction(self.action_Add_API_URL)
        self.menu_Query.addAction(self.action_New_Query)
        self.menu_Query.addSeparator()
        self.menu_Query.addAction(self.actionSave_Query_Results)
        self.menubar.addAction(self.menu_Open.menuAction())
        self.menubar.addAction(self.menu_Query.menuAction())
        self.menubar.addAction(self.menu_About.menuAction())

        self.retranslateUi(MainWindow)
        QtCore.QMetaObject.connectSlotsByName(MainWindow)

    def retranslateUi(self, MainWindow):
        _translate = QtCore.QCoreApplication.translate
        MainWindow.setWindowTitle(_translate("MainWindow", "MainWindow"))
        self.Query.setText(_translate("MainWindow", "Query"))
        self.razorpayUPI.setText(_translate("MainWindow", "Use RazorPay (Requires API Key)"))
        self.upibankAPI.setText(_translate("MainWindow", "Use UPIBankValidator"))
        self.label_4.setText(_translate("MainWindow", "Email Address"))
        self.pasteEmail.setText(_translate("MainWindow", "Paste"))
        self.label.setText(_translate("MainWindow", "Name:"))
        self.label_2.setText(_translate("MainWindow", "Phone No."))
        self.allUPI.setText(_translate("MainWindow", "All UPI"))
        self.mobileUPI.setText(_translate("MainWindow", "Mobile UPI"))
        self.gpayUPI.setText(_translate("MainWindow", "Gpay UPI"))
        self.saveButton.setText(_translate("MainWindow", "Save"))
        self.pastePhone.setText(_translate("MainWindow", "Paste"))
        self.footer.setText(_translate("MainWindow", "Search Query"))
        self.fastagUPI.setText(_translate("MainWindow", "Fastag UPI"))
        self.customAPI.setText(_translate("MainWindow", "Use Custom API"))
        self.menu_Open.setTitle(_translate("MainWindow", "&API Key"))
        self.menu_Add_API_2.setTitle(_translate("MainWindow", "&Add API Key"))
        self.menu_Query.setTitle(_translate("MainWindow", "&Query"))
        self.menu_About.setTitle(_translate("MainWindow", "&About"))
        self.actionManually.setText(_translate("MainWindow", "Manually"))
        self.action_Manually.setText(_translate("MainWindow", "&Manually"))
        self.action_From_File.setText(_translate("MainWindow", "&From File"))
        self.action_New_Query.setText(_translate("MainWindow", "&New Query"))
        self.actionSave_Query_Results.setText(_translate("MainWindow", "Save Query Results"))
        self.action_Add_API_URL.setText(_translate("MainWindow", "& Add API URL"))


gupi_recon/__main__.py
import sys
import requests
import pyperclip
import concurrent.futures

from sys import exit
from time import sleep
from pathlib import Path
from random import uniform as rand

from PyQt5.QtCore import QRegExp, QObject, QThread, pyqtSignal
from PyQt5.QtGui import QRegExpValidator, QIcon
from PyQt5.QtWidgets import (
    QApplication, QDialog, QMainWindow, QDialogButtonBox, QLineEdit,
    QTableWidgetItem, QFileDialog
)
from PyQt5 import QtWidgets
from PyQt5.uic import loadUi

from .ui import Ui_MainWindow


API_URL = 'https://upibankvalidator.com/api/upiValidation?upi='

phone_regex = QRegExp("[0-9]{10}")
email_regex = QRegExp("^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\\.[a-zA-Z]{2,4}$")

#  opting to load lists from a file instead of hardcoding them
#  as this would be more flexible, allow for easier updates,
#  and allow others to make use of the lists provided

# Load module resources by finding path.
RESOURCE_PATH = (
    f'{Path(sys.modules["gupi_recon"].__file__).parent}/data'
    if 'gupi_recon' in sys.modules else
    'gupi_recon/data'
)

with open(f"{RESOURCE_PATH}/general_suffixes.txt", "r") as suffix_file:
    upi_suffix_dict = suffix_file.read().splitlines()  # read all suffixes into a list

with open(f"{RESOURCE_PATH}/mobile_suffixes.txt", "r") as mobile_suffix_file:
    mobile_suffix_dict = mobile_suffix_file.read().splitlines()

with open(f"{RESOURCE_PATH}/fastag_suffixes.txt", "r") as fastag_suffix_file:
    fastag_suffix_dict = fastag_suffix_file.read().splitlines()

with open(f"{RESOURCE_PATH}/gpay_suffixes.txt", "r") as gpay_suffix_file:
    gpay_suffix_dict = gpay_suffix_file.read().splitlines()


class Scraper(QObject):
    finished = pyqtSignal()
    progress = pyqtSignal(int)
    found = 0
    count = 0

    def searchvpa(self, searchtext, vpa_dict, threadcount):
        if (threadcount == 0):
            for suffix in vpa_dict:
                try:
                    self.address_discovery(f'{searchtext}@{suffix}', API_URL)
                except KeyboardInterrupt:
                    exit(0)
        else:
            threadcount = min(threadcount, 10)
            with concurrent.futures.ThreadPoolExecutor(max_workers=threadcount) as executor:
                try:
                    for suffix in vpa_dict:
                        executor.submit(self.address_discovery,
                                        f'{searchtext}@{suffix}', API_URL)
                        sleep(rand(.1, .2))
                except KeyboardInterrupt:
                    #  quit ungracefully on keyboard interrupt:
                    #  considering the bandwidth consumed for requests,
                    #  there is no reason to wait for the threads to finish
                    #  sorry for the inconvenience
                    executor._threads.clear()
                    concurrent.futures.thread._threads_queues.clear()
        if self.found == 0:
            win.set_footer(f"No VPA found for: {searchtext}", "red")

    def address_discovery(self, vpa, api_url):
        r = requests.post(api_url + vpa)

        print(r.status_code)
        if r.status_code == 200 and r.json()['isUpiRegistered'] is True:
            rowPosition = win.Result.rowCount()
            win.Result.insertRow(rowPosition)
            win.Result.setItem(
                rowPosition, 0, QTableWidgetItem(r.json()['name']))
            win.Result.setItem(rowPosition, 1, QTableWidgetItem(vpa))
            win.nameEdit.setText(r.json()['name'])
            self.found += 1
            win.set_footer(f"Found: {self.found}", "green")
        if r.status_code == 400:
            win.set_footer("Bad Request", "red")
        if r.status_code == 429:
            win.set_footer("Too Many Requests", "red")
        self.progress.emit(self.count)
        self.count += 1

    def run(self):
        self.searchvpa(win.phoneEdit.text(), win.suffix, 2)
        self.finished.emit()


class Window(QMainWindow, Ui_MainWindow):
    def __init__(self, parent=None):
        super().__init__(parent)
        self.setupUi(self)
        self.connectSignalsSlots()
        self.phoneEdit.setValidator(
            QRegExpValidator(phone_regex, self.phoneEdit))
        self.emailEdit.setValidator(
            QRegExpValidator(email_regex, self.emailEdit))
        self.nameEdit.setReadOnly(True)
        self.progressBar.setVisible(False)
        self.allUPI.setChecked(True)
        self.upibankAPI.setChecked(True)
        self.Result.setColumnCount(2)
        self.Result.horizontalHeader().setSectionResizeMode(QtWidgets.QHeaderView.Stretch)
        self.Result.setHorizontalHeaderLabels(["Name", "VPA"])
        self.setWindowIcon(QIcon('logo.png'))
        self.setWindowTitle("gUPI recon")

    def connectSignalsSlots(self):
        self.action_Manually.triggered.connect(self.add_api)
        self.pastePhone.clicked.connect(self.paste_phone)
        self.pasteEmail.clicked.connect(self.paste_email)
        self.Query.clicked.connect(self.query)
        self.saveButton.clicked.connect(self.saveQuery)
        self.action_Add_API_URL.triggered.connect(self.add_api)
        # self.action_Find_Replace.triggered.connect(self.findAndReplace)
        # self.action_About.triggered.connect(self.about)

    def add_api(self):
        pass

    def set_footer(self, text, color="black"):
        self.footer.setStyleSheet(f"color: {color}")
        self.footer.setText(text)

    def add_api(self):
        dialog = add_API(self)
        dialog.exec()

    def paste_phone(self):
        if (QRegExp.exactMatch(phone_regex, pyperclip.paste())):
            self.phoneEdit.setText(pyperclip.paste())
        else:
            self.set_footer("Invalid phone number", "red")

    def paste_email(self):
        if (QRegExp.exactMatch(email_regex, pyperclip.paste())):
            self.emailEdit.setText(pyperclip.paste())
        else:
            self.set_footer("Invalid email", "red")

    def reportProgress(self, value):
        self.progressBar.setValue(round(((value + 1)/len(self.suffix)) * 100))

    def saveQuery(self):
        if self.Result.rowCount() > 0:
            name, _ = QFileDialog.getSaveFileName(
                self, 'Save Query', self.phoneEdit.text(), 'CSV(*.csv)')
            if not name:
                return
            with open(name, 'w') as file:
                for i in range(self.Result.rowCount()):
                    file.write(self.Result.item(i, 0).text())
                    file.write(", ")
                    file.write(self.Result.item(i, 1).text())
                    file.write("\n")
            footer_text = f"Saved to {name}"
            if len(footer_text) > 27:
                self.set_footer(f"{footer_text[25:]}...", "green")
            self.set_footer(footer_text, "green")
        else:
            self.set_footer("No results to save", "red")

    def query(self):
        # set footer
        self.set_footer("Querying...")
        self.thread = QThread()
        self.worker = Scraper()
        self.worker.moveToThread(self.thread)
        self.thread.started.connect(self.worker.run)
        self.worker.finished.connect(self.thread.quit)
        self.worker.finished.connect(self.worker.deleteLater)
        self.thread.finished.connect(self.thread.deleteLater)
        self.worker.progress.connect(self.reportProgress)
        self.progressBar.setValue(0)
        self.progressBar.setVisible(True)
        self.thread.finished.connect(
            lambda: self.progressBar.setVisible(False)
        )
        self.suffix = upi_suffix_dict
        if (win.gpayUPI.isChecked()):
            self.suffix = gpay_suffix_dict
        elif (win.fastagUPI.isChecked()):
            self.suffix = fastag_suffix_dict
        elif (win.mobileUPI.isChecked()):
            self.suffix = mobile_suffix_dict
        self.Result.clearContents()
        self.Result.setRowCount(0)
        self.thread.start()


class add_API(QDialog):
    def ok_clicked(self):
        print(self.api_input.text())
        self.close()

    def __init__(self, parent=None):
        super().__init__(parent)
        loadUi("ui/api_input.ui", self)
        self.api_input = self.findChild(QLineEdit, "lineEdit")
        self.button_box = self.findChild(QDialogButtonBox, "buttonBox")
        self.ok = self.button_box.buttons()[0]
        self.ok.clicked.connect(self.ok_clicked)


if __name__ == "__main__":
    app = QApplication(sys.argv)
    win = Window()
    win.show()
    sys.exit(app.exec())
